### PR TITLE
ci: k8s: adapt gha-run.sh to run locally

### PR DESCRIPTION
**Pull request by wainersm**
My first version of https://github.com/kata-containers/kata-containers/pull/7600 I simply didn't test locally, instead I let the CI run. Unfortunately CI failed and it took a good amount of time to give me a feedback. Other than that, I don't need the kata-deploy payload rebuilt as I am just changing tests files. Ok, so I decided to look if I could make `gha-run.sh` (which basically implements the CI workflow) to run locally, and here is this PR. I hope it helps to speed up the process to migrate testing code from CCv0 -> main.

I chose to deploy and managed a k8s locally with kcli (https://github.com/karmab/kcli) because it is a tool that I use almost everyday (so I know it quite well). I know that @fidencio used to use it, and I suspect others on the community use too.

kcli will be running with one control node and one worker. Two tests are failing on this configuration because they expect a single-node cluster. I fixed both tests. 

Here is how I run the tests:

```
$ ./gha-run.sh create-cluster-kcli
$ ./gha-run.sh deploy-kata-kcli
$ ./gha-run.sh run-tests-kcli
```

If I need to re-deploy kata for some reason I run `./gha-run.sh cleanup-kcli && ./gha-run.sh deploy-kata-kcli`. Finally I can delete the cluster entirely with a `./gha-run.sh delete-cluster-kcli`

Fixes https://github.com/kata-containers/kata-containers/issues/7620

### CODE REVIEW

File: tests/integration/kubernetes/gha-run.sh
Comment (fidencio) [inline tests/integration/kubernetes/gha-run.sh:66]: I'd also add `-P cpumodel=host-passthrough`, as this will avoid some issues when testing Cloud Hypervisor.

**Review by fidencio — COMMENTED** [review]
@wainersm, I didn't test it locally yet, but I will do after you rebase and solve the conflict we're having.

In general it looks very good, I just had one comment so we can Cloud Hypervisor to also work as part of this.

**Comment by sprt** [comment]
Love this! Can we get docs in a readme on how to run this?

**Comment by wainersm** [comment]
Finally resolved the conflicts. Re-ran the tests locally, all passed:

```
[wmoschet@wainer-laptop kubernetes]$ ./gha-run.sh run-tests-kcli
namespace/kata-containers-k8s-tests created
Context "kubernetes-admin@kubernetes" modified.
~/go/src/github.com/kata-containers/kata-containers/tests/integration/kubernetes ~/go/src/github.com/kata-containers/kata-containers/tests/integration/kubernetes
[run_kubernetes_tests.sh:161] INFO: Policy tests are disabled on this platform
[run_kubernetes_tests.sh:137] INFO: Test actions that must be successful
k8s-confidential.bats
 - Test unencrypted confidential container launch success and verify that we are running in a secure enclave. (skipped: Test not supported for qemu.)

1 test, 0 failures, 1 skipped

k8s-attach-handlers.bats
 ✓ Running with postStart and preStop handlers

1 test, 0 failures

k8s-caps.bats
 ✓ Check capabilities of pod

1 test, 0 failures

k8s-configmap.bats
 ✓ ConfigMap for a pod

1 test, 0 failures

k8s-copy-file.bats
 ✓ Copy file in a pod
 ✓ Copy from pod to host

2 tests, 0 failures

k8s-cpu-ns.bats
 ✓ Check CPU constraints

1 test, 0 failures

k8s-credentials-secrets.bats
 ✓ Credentials using secrets

1 test, 0 failures

k8s-custom-dns.bats
 ✓ Check custom dns

1 test, 0 failures

k8s-empty-dirs.bats
 ✓ Empty dir volumes
 ✓ Empty dir volume when FSGroup is specified with non-root container

2 tests, 0 failures

k8s-env.bats
 ✓ Environment variables

1 test, 0 failures

k8s-exec.bats
 ✓ Kubectl exec

1 test, 0 failures

k8s-file-volume.bats
 ✓ Test readonly volume for pods

1 test, 0 failures

k8s-inotify.bats
 ✓ configmap update works, and preserves symlinks

1 test, 0 failures

k8s-job.bats
 ✓ Run a job to completion

1 test, 0 failures

k8s-kill-all-process-in-container.bats
 ✓ Kill all processes in container

1 test, 0 failures

k8s-limit-range.bats
 ✓ Limit range for storage

1 test, 0 failures

k8s-liveness-probes.bats
 ✓ Liveness probe
 ✓ Liveness http probe
 ✓ Liveness tcp probe

3 tests, 0 failures

k8s-memory.bats
 ✓ Exceeding memory constraints
 ✓ Running within memory constraints

2 tests, 0 failures

k8s-nested-configmap-secret.bats
 ✓ Nested mount of a secret volume in a configmap volume for a pod

1 test, 0 failures

k8s-oom.bats
 ✓ Test OOM events for pods

1 test, 0 failures

k8s-optional-empty-configmap.bats
 ✓ Optional and Empty ConfigMap Volume for a pod

1 test, 0 failures

k8s-optional-empty-secret.bats
 ✓ Optional and Empty Secret Volume for a pod

1 test, 0 failures

k8s-pid-ns.bats
 ✓ Check PID namespaces

1 test, 0 failures

k8s-pod-quota.bats
 ✓ Pod quota

1 test, 0 failures

k8s-port-forward.bats
 - Port forwarding (skipped: test not working see: https://github.com/kata-containers/runtime/issues/1834)

1 test, 0 failures, 1 skipped

k8s-projected-volume.bats
 ✓ Projected volume

1 test, 0 failures

k8s-qos-pods.bats
 ✓ Guaranteed QoS
 ✓ Burstable QoS
 ✓ BestEffort QoS

3 tests, 0 failures

k8s-replication.bats
 ✓ Replication controller

1 test, 0 failures

k8s-seccomp.bats
 ✓ Support seccomp runtime/default profile

1 test, 0 failures

k8s-sysctls.bats
 ✓ Setting sysctl

1 test, 0 failures

k8s-security-context.bats
 ✓ Security context

1 test, 0 failures

k8s-shared-volume.bats
 ✓ Containers with shared volume
 ✓ initContainer with shared volume

2 tests, 0 failures

k8s-volume.bats
 ✓ Create Persistent Volume

1 test, 0 failures

k8s-nginx-connectivity.bats
 ✓ Verify nginx connectivity between pods

1 test, 0 failures

~/go/src/github.com/kata-containers/kata-containers/tests/integration/kubernetes
```

**Comment by wainersm** [comment]
hi @sprt ! can I have your review please?

File: tests/integration/kubernetes/gha-run.sh
Comment (sprt) [inline tests/integration/kubernetes/gha-run.sh:161]: I would define a default `KCLI_CLUSTER_NAME` at the top level so that we don't repeat this `kata-k8s` string all over. Or even put this export in a function.

**Review by sprt — APPROVED** [review]
Thanks for the meticulous change, just left a nit!

Also can we document this somewhere? I'm sure folks would find this useful. Literally just what you wrote in the PR description would be enough.

**Comment by wainersm** [comment]
hi @sprt ! Thanks for your review! @fidencio will start a CI guide, I am planning to write about how to run this locally on that documento. Is it okay to you?

@fidencio can I have your approval (or not) to merge?

**Comment by fidencio** [comment]
So, I've spent some time testing it Today, it works as described, but I do think we need to expand a little bit on how things are done with kcli, and this is **NOT** material for this PR, but rather follow-ups to improve it.

With that said, here's the list of enhancements I'd like to see happening:
* `./gha-run.sh create-cluster-kcli` must be expanded to create the cluster type matching with the clusters we're testing
  * kcli supports creating: aks, generic, and k3s clusters
    * https://github.com/kata-containers/kata-containers/issues/8128
    * https://github.com/kata-containers/kata-containers/issues/8129
  * kcli can be expanded to also create: rke2, and k0s clusters
    * This not necessarily on us to do, but it's good to keep track of that
      * https://github.com/kata-containers/kata-containers/issues/8130
      * https://github.com/kata-containers/kata-containers/issues/8131
*  `./gha-run.sh create-cluster-kcli`  must be expanded to match "smaller" / "normal" VM sizes and only run the tests accordingly
  * This is basically related to ensuring the amount of CPU / RAM is properly allocated accordingly to what we do with the instances we use
  * I think this can simply be done tweaking the params of the cluster created
  * https://github.com/kata-containers/kata-containers/issues/8132
*  `./gha-run.sh create-cluster-kcli` must ensure that the cluster created will work with all the hypervisors
  * Cloud Hypervisor requires `-P cpumodel=host-passthrough` passed to the VM, otherwise it won't work
  * https://github.com/kata-containers/kata-containers/issues/8133

* `./gha-run.sh deploy-kata-kcli` *must* respect all the environment variables passed, which is something I didn't test but I would like to hear from you, @wainersm, whether that's the case or not
  * This is needed in order to ensure that the kata-deploy will use exactly the same env-vars used as part of the tests
  * I didn´t open an issue for this one, and I'd like to ask you to open in case this is not respected
* `./gha-run.sh deploy-kata-kcli` could take a directory as an argument and just apply those
  * What I have in mind here is to upload the kata-deploy folder as part of the run, then we can point users to just download that folder, and run `./gha-run.sh deploy-kata-kcli $folder_path`, which will ensure that the test will be replicated with the exact same parameters as the CI
  * https://github.com/kata-containers/kata-containers/issues/8134
  * https://github.com/kata-containers/kata-containers/issues/8135
  * https://github.com/kata-containers/kata-containers/issues/8136
*  `./gha-run.sh deploy-kata-kcli` by default should create **ALL** the runtime classes, unless specified the opposite
  * The current behaviour is to create only the QEMU one
  * https://github.com/kata-containers/kata-containers/issues/8137
*  `./gha-run.sh deploy-kata-kcli` could take a PR as an argument
  * The tag of the image used is not that hard to build, as it basically is: $pr_number-$sha_of_the_last_commit-$arch -- and that information can be easily found in a programatically way
  * https://github.com/kata-containers/kata-containers/issues/8138

* `./gha-run.sh run-tests-kcli` should respect the instance size that's being used
  * https://github.com/kata-containers/kata-containers/issues/8139
* `./gha-run.sh run-tests-kcli` could take a test as an argument, to speed up the debugging process
  * https://github.com/kata-containers/kata-containers/issues/8140

I've taken the freedom to go ahead and open issues for the majority of those (apart from one that is waiting on your answer), and I really expect to see the follow-up PRs coming, but this one is good to go as it is.

File: tests/integration/kubernetes/k8s-file-volume.bats
Comment (fidencio) [inline tests/integration/kubernetes/k8s-file-volume.bats:31]: Nit: wrong identation.

**Review by fidencio — APPROVED** [review]
lgtm, with the caveat that the issues opened must be followed-up as soon as this PR gets merged.

**Comment by wainersm** [comment]
> So, I've spent some time testing it Today, it works as described, but I do think we need to expand a little bit on how things are done with kcli, and this is **NOT** material for this PR, but rather follow-ups to improve it.
> 
> With that said, here's the list of enhancements I'd like to see happening:
> 
>     * `./gha-run.sh create-cluster-kcli` must be expanded to create the cluster type matching with the clusters we're testing
>       
>       * kcli supports creating: aks, generic, and k3s clusters
>         
>         * [ci: local: `./gha-run.sh create-cluster-kcli` must be expanded to support creating AKS clusters #8128](https://github.com/kata-containers/kata-containers/issues/8128)
>         * [ci: local: ./gha-run.sh create-cluster-kcli must be expanded to support creating k3s clusters #8129](https://github.com/kata-containers/kata-containers/issues/8129)
>       * kcli can be expanded to also create: rke2, and k0s clusters
>         
>         * This not necessarily on us to do, but it's good to keep track of that
>           
>           * [ci: local: ./gha-run.sh create-cluster-kcli must be expanded to support creating rke2 clusters #8130](https://github.com/kata-containers/kata-containers/issues/8130)
>           * [ci: local: ./gha-run.sh create-cluster-kcli must be expanded to support creating k0s clusters #8131](https://github.com/kata-containers/kata-containers/issues/8131)
>

I confess  I had hidden plans to leverage kcli to deploy openShift on CI but I wasn't thinking in expand to create other types ks0, rancher...etc... I think it is a good idea, having one tool to rule them all (or almost all). I will be working on that.

>     * `./gha-run.sh create-cluster-kcli`  must be expanded to match "smaller" / "normal" VM sizes and only run the tests accordingly
> 
>     * This is basically related to ensuring the amount of CPU / RAM is properly allocated accordingly to what we do with the instances we use
> 
>     * I think this can simply be done tweaking the params of the cluster created
> 
>     * [ci: local: ./gha-run.sh create-cluster-kcli must be expanded to support creating "smaller" / "normal" instance sizes #8132](https://github.com/kata-containers/kata-containers/issues/8132)
> 

Makes sense. I have some questions but better move the discussion to the issue.

>     * `./gha-run.sh create-cluster-kcli` must ensure that the cluster created will work with all the hypervisors
> 
>     * Cloud Hypervisor requires `-P cpumodel=host-passthrough` passed to the VM, otherwise it won't work
> 
>     * [ci: local: ./gha-run.sh create-cluster-kcli must create VMs passing  `-P cpumodel=host-passthrough` #8133](https://github.com/kata-containers/kata-containers/issues/8133)
> 

Low hanging fruit, will send on the follow up.

>     * `./gha-run.sh deploy-kata-kcli` _must_ respect all the environment variables passed, which is something I didn't test but I would like to hear from you, @wainersm, whether that's the case or not
>       
>       * This is needed in order to ensure that the kata-deploy will use exactly the same env-vars used as part of the tests
>       * I didn´t open an issue for this one, and I'd like to ask you to open in case this is not respected
> 
>     * `./gha-run.sh deploy-kata-kcli` could take a directory as an argument and just apply those
>       
>       * What I have in mind here is to upload the kata-deploy folder as part of the run, then we can point users to just download that folder, and run `./gha-run.sh deploy-kata-kcli $folder_path`, which will ensure that the test will be replicated with the exact same parameters as the CI
>       * [ci: local: kata-deploy directory could be uploaded during the runs, so users could simply download and use it in their local test #8134](https://github.com/kata-containers/kata-containers/issues/8134)
>       * [ci: local: `./gha-run.sh deploy-kata-kcli` could take a directory as an argument #8135](https://github.com/kata-containers/kata-containers/issues/8135)
>       * [ci: local: `./gha-run.sh deploy-kata-kcli` could take the k8s flavour as an argument #8136](https://github.com/kata-containers/kata-containers/issues/8136)
> 

I never thought about it and is a great idea. For what I've observer, kata devels find difficult to figure out the parameters to re-run the tests locally. The Jenkins-based CI we had at least a single environment variable (`CI_JOB`) used to source all the others, with the new CI we don't have this and users should simply read the workflows yamls.

This remind me of Avocado's replay feature (https://avocado-framework.readthedocs.io/en/latest/quickstart/index.html?highlight=replay#job-replay-and-job-diff).


>     * `./gha-run.sh deploy-kata-kcli` by default should create **ALL** the runtime classes, unless specified the opposite
> 
>     * The current behaviour is to create only the QEMU one
> 
>     * [ci: local: `./gha-run.sh deploy-kata-kcli` should create all the runtime classes by default #8137](https://github.com/kata-containers/kata-containers/issues/8137)
> 
>     * `./gha-run.sh deploy-kata-kcli` could take a PR as an argument
> 
>     * The tag of the image used is not that hard to build, as it basically is: $pr_number-$sha_of_the_last_commit-$arch -- and that information can be easily found in a programatically way
> 
>     * [ci: local: `./gha-run.sh deploy-kata-kcli` could take a PR as an argument #8138](https://github.com/kata-containers/kata-containers/issues/8138)
> 

Yep, another difficult that I've observed is about figuring out the image....

Ah, some workflows relies on the kata tarball. Even if we are archiving it and the user can simply download, I think we should document how it is created; the tarball is useful for development too...  https://github.com/kata-containers/kata-containers/issues/8181

>     * `./gha-run.sh run-tests-kcli` should respect the instance size that's being used
>       
>       * [ci: local: `./gha-run.sh run-tests-kcli` should respect the instance size that's being used #8139](https://github.com/kata-containers/kata-containers/issues/8139)
> 
>     * `./gha-run.sh run-tests-kcli` could take a test as an argument, to speed up the debugging process
>       
>       * [ci: local:  `./gha-run.sh run-tests-kcli` could take a test as an argument #8140](https://github.com/kata-containers/kata-containers/issues/8140)
> 
> 
> I've taken the freedom to go ahead and open issues for the majority of those (apart from one that is waiting on your answer), and I really expect to see the follow-up PRs coming, but this one is good to go as it is.

Thanks for that, sound as a good plan of improvements!

**Comment by wainersm** [comment]
Fixed a small indentation error and rebased so the CI jobs get triggered.

**Pull request merged**
